src/array/dictionary/value_map.rs

```diff
@@ -91,24 +141,22 @@ impl<K: DictionaryKey, M: MutableArray> ValueMap<K, M> {
         V: AsIndexed<M>,
         M::Type: Eq + Hash,
     {
-        let mut hasher = self.map.hasher().build_hasher();
-        value.as_indexed().hash(&mut hasher);
-        let hash = hasher.finish();
-
+        let hash = ahash_hash(value.as_indexed());
         Ok(
-            match self.map.raw_entry_mut().from_hash(hash, |key| {
+            match self.map.raw_entry_mut().from_hash(hash, |item| {
                 // safety: we've already checked (the inverse) when we pushed it, so it should be ok?
-                let index = unsafe { key.as_usize() };
+                let index = unsafe { item.key.as_usize() };
                 // safety: invariant of the struct, it's always in bounds since we maintain it
                 let stored_value = unsafe { self.values.value_unchecked_at(index) };
                 stored_value.borrow() == value.as_indexed()
             }) {
-                RawEntryMut::Occupied(entry) => *entry.key(),
+                RawEntryMut::Occupied(entry) => entry.key().key,
                 RawEntryMut::Vacant(entry) => {
                     let index = self.values.len();
                     let key = K::try_from(index).map_err(|_| Error::Overflow)?;
-                    entry.insert_hashed_nocheck(hash, key, ()); // NB: don't use .insert() here!
+                    entry.insert_hashed_nocheck(hash, Hashed { hash, key }, ()); // NB: don't use .insert() here!
                     push(&mut self.values, value)?;
+                    assert_eq!(self.values.len(), index + 1);
```

Comment:
Shall we make this a `debug_assert`? Don't see this going wrong.

Reply:
Good catch; I think this was actually debug_assert_eq in one of the previous iterations (and was meant to be ) but got messed during all the rewrites.